backends/dpdk/dpdkHelpers.cpp

```diff
@@ -134,27 +135,64 @@ bool ConvertStatementToDpdk::preorder(const IR::AssignmentStatement *a) {
                 src1Op = src1Member;
             }
         }
+        /* If a constant is present in binary operation and it is not 8-bit
+           aligned and it is signed, then the constant value is sign extended
+           with a width of either 32 bit if its original bit-width is less than
+           32 bit or 64 bit if the bit-width of constant is greater than 32 bit
+        */
+
+        if (!isEightBitAligned(src2Op)) {
+            if (auto tb = src2Op->type->to<IR::Type_Bits>()) {
+                if (tb != nullptr && tb->isSigned) {
+                    isSignExtended = true;
+                    auto consOrgBitwidth = src2Op->to<IR::Expression>()->type->width_bits();
+                    auto bitWidth = consOrgBitwidth < 32 ? 32 : 64;
+                    if (src2Op->is<IR::Constant>()) {
+                        auto consValue = src2Op->to<IR::Constant>()->value;
+                        auto val = consValue << (bitWidth - consOrgBitwidth);
+                        consValue = val >> (bitWidth - consOrgBitwidth);
+                        src2Op = new IR::Constant(IR::Type_Bits::get(bitWidth), consValue);
+                    } else {
+                        auto val = bitWidth - consOrgBitwidth;
+                        add_instr(new IR::DpdkShlStatement(src2Op, src2Op, new IR::Constant(val)));
```

Comment:
so this is a sign-extension?
some languages have two kinds of shift right: arithmetic and logic. I guess you need arithmetic.
Is the dpdk shift right semantics based on the type?

Reply:
Shift operations are logical in DPDK.
For sign extension of non-constant variables, i am moving the msb of variable through left and right operation.
Adding snippet for the same ( sign extended 16 bit to 32 bit )
(Snip)
    #include <iostream>
#include <cstdint> 
using namespace std;

int main() {
    int16_t var;
    int b = 17;
    cin>>var;
    int32_t var2 = var << 16;
   cout<<"Size of var "<<sizeof(var)<<endl;
    cout<<hex <<var<< dec <<endl;
    

    var2 = var2 >> 16;
    cout<<"Size of var2 "<<sizeof(var2)<<endl;
    cout<<hex <<var2<< dec <<endl;
    return 0;
}

Input : -2
Output :
Size of var 2
fffe
Size of var2 4
fffffffe
(Snip)